File: src/main/java/com/rubensminoru/main/Processor.java
package com.rubensminoru.main;

import com.rubensminoru.messages.KafkaMessage;
import com.rubensminoru.partitioners.Partitioner;
import com.rubensminoru.writers.Writer;
import com.rubensminoru.writers.WriterFactory;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Processor {
    private String topic;
    private final WriterFactory writerFactory;
    private final Partitioner partitioner;

    private Map<Writer, List<TopicInfo>> writerTopicInfos;
    private List<TopicInfo> topicInfos;

    public Processor(String topic, WriterFactory writerFactory, Partitioner partitioner) {
        this.topic = topic;
        this.writerFactory = writerFactory;
        this.partitioner = partitioner;
        this.writerTopicInfos = new HashMap();
        this.topicInfos = new ArrayList<>();
    }

    public boolean process(List<KafkaMessage> messages) {
        for (KafkaMessage message : messages) {
            process(message);
        }

        return true;
    }

    public void process(KafkaMessage message) {
        this.writerTopicInfos = this.addOrUpdateWriter(this.writerTopicInfos, message);

        this.topicInfos = this.addOrUpdatePartitionInfo(this.topicInfos, message);
    }

    public Map<Writer, List<TopicInfo>> addOrUpdateWriter(Map<Writer, List<TopicInfo>> localWriterTopicInfos, KafkaMessage message) {
        Writer currentWriter = null;
        List<TopicInfo> localTopicInfos = new ArrayList<>();

        for (Writer writer : localWriterTopicInfos.keySet()) {
            if (checkRecordPartition(message, writer)) {
                currentWriter = writer;
            }
        }

        if (currentWriter == null) {
            currentWriter = this.writerFactory.createInstance();
        } else {
            localTopicInfos = localWriterTopicInfos.get(currentWriter);
        }

        localTopicInfos = this.addOrUpdatePartitionInfo(localTopicInfos, message);

        currentWriter.write(message);

        localWriterTopicInfos.put(currentWriter, localTopicInfos);

        return localWriterTopicInfos;
    }

    public List<TopicInfo> addOrUpdatePartitionInfo(List<TopicInfo> localTopicInfos, KafkaMessage message) {
        Integer partitionIndex = null;
        TopicInfo topicInfo;

        for (TopicInfo localTopicInfo : localTopicInfos) {
            if (localTopicInfo.getPartition() == message.getPartition()) {
                partitionIndex = localTopicInfos.indexOf(localTopicInfo);
            }
        }

        if (partitionIndex == null) {
            topicInfo = new TopicInfo(this.topic, message.getPartition(), message.getOffset());
            localTopicInfos.add(topicInfo);
        } else {
            topicInfo = localTopicInfos.get(partitionIndex);

            topicInfo.setOffset(message.getOffset());

            localTopicInfos.set(partitionIndex, topicInfo);
        }

        return localTopicInfos;
    }

    public boolean checkRecordPartition(KafkaMessage message, Writer writer) {
        return partitioner.check(message, writer);
    }

    public static class TopicInfo {
        private final String topic;
        private final int partition;
        private long offset;

        public TopicInfo(String topic, int partition, long offset) {
            this.topic = topic;
            this.partition = partition;
            this.offset = offset;
        }

        public void setOffset(long offset) {
            this.offset = offset;
        }

        public int getPartition() {
            return partition;
        }

        public String getTopic() { return topic; }

        public long getOffset() { return offset; }
    }
}

File: src/main/scala/com/rubensminoru/Processor.scala
package com.rubensminoru;

import com.rubensminoru.messages.KafkaAvroMessage
import com.rubensminoru.partitioners.Partitioner
import com.rubensminoru.writers.Writer
import com.rubensminoru.writers.FileWriterFactory
import java.io.IOException

import org.apache.kafka.clients.consumer.ConsumerRecord

class Processor(topic:String, path:String, partitioner:Partitioner) {
    // {writer => partition, topicinfo}    
    var writerTopicInfos:Map[Writer, Map[Int, TopicInfo]] = Map()
    var topicInfos:Map[Int, TopicInfo] = Map()

    @throws(classOf[IOException])
    def process(messages:Iterable[KafkaAvroMessage]):Boolean = {
        for (message <- messages) {
            processEach(message)
        }

        true
    }

    @throws(classOf[IOException])
    def processEach(message:KafkaAvroMessage) = {
        this.writerTopicInfos = addOrUpdateWriter(this.writerTopicInfos, message);

        this.topicInfos = addOrUpdatePartitionInfo(this.topicInfos, message);
    }

    def addOrUpdateWriter(localWriterTopicInfos:Map[Writer, Map[Int, TopicInfo]], message:KafkaAvroMessage):Map[Writer, Map[Int, TopicInfo]] = {
        val localWriterTopicInfo = localWriterTopicInfos
          .find(kv => checkRecordPartition(message, kv._1))
          .get

        if(localWriterTopicInfo == null) {
            val currentWriter = FileWriterFactory.apply(message.schema, this.path)
            localWriterTopicInfos + (currentWriter -> this.addOrUpdatePartitionInfo(Map[Int, TopicInfo](), message))
        } else {
            localWriterTopicInfos + (localWriterTopicInfo._1 -> this.addOrUpdatePartitionInfo(localWriterTopicInfo._2, message))
        }
    }

    def addOrUpdatePartitionInfo(localTopicInfos:Map[Int, TopicInfo], message:KafkaAvroMessage): Map[Int, TopicInfo] = {
        val topicInfo = localTopicInfos.getOrElse(message.partition, new TopicInfo(this.topic, message.partition, message.offset))
        topicInfo.offset = message.offset

        localTopicInfos + (message.partition -> topicInfo)
    }

    def checkRecordPartition(message:KafkaAvroMessage, writer:Writer): Boolean = partitioner.check(message, writer)
}

class TopicInfo(var topic:String, var partition:Int, var offset:Long)